### PR TITLE
GEODE-4678: Remove ServerHandshakeProcessor and refactor handshake creation

### PR DESCRIPTION
**Pull request by WireBaron**
Note that this change is currently provided as several distinct commits, it will likely be easier to follow the refactoring stepping through the commits individually.  The two commits which should be reviewed a bit more closely (are more than just simple refactoring) are:

b579d04 - Changing ClientHealthMonitor cleanup maps to use typenames ...
132c690 - Refactored ServerSideHandshakeFactory to more cleanly separate the Ha… 

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerSideHandshakeImpl.java [hunk 1]
```diff
@@ -53,63 +60,55 @@ public ServerSideHandshakeImpl(Socket sock, int timeout, DistributedSystem sys,
...
+          // versioned streams allow object serialization code to deal with older clients
+          dataInputStream = new VersionedDataInputStream(dataInputStream, clientVersion);
+          dataOutputStream = new VersionedDataOutputStream(dataOutputStream, clientVersion);
         }
-        this.replyCode = (byte) valRead;
-        if (this.replyCode != REPLY_OK) {
-          throw new IOException(
-              LocalizedStrings.HandShake_HANDSHAKE_REPLY_CODE_IS_NOT_OK.toLocalizedString());
+        this.id = ClientProxyMembershipID.readCanonicalized(dataInputStream);
+        // Note: credentials should always be the last piece in handshake for
```
Comment (PivotalSarge) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerSideHandshakeImpl.java]: Nice commentary. Future You will thank you.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerSideHandshakeImpl.java [hunk 2]
```diff
@@ -53,63 +60,55 @@ public ServerSideHandshakeImpl(Socket sock, int timeout, DistributedSystem sys,
     this.securityService = securityService;
     this.encryptor = new EncryptorImpl(sys.getSecurityLogWriter());
 
-    {
-      int soTimeout = -1;
+    int soTimeout = -1;
```
Comment (bschuchardt) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerSideHandshakeImpl.java]: I can't tell what changed here - I assume it's github's diff going crazy

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java [hunk 3]
```diff
@@ -671,22 +662,6 @@ public boolean isClientServerConnection() {
     return communicationMode.isClientToServerOrSubscriptionFeed();
   }
 
-  static class Counter {
```
Comment (PivotalSarge) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java]: How was using `Counter` any clearer than `int`? Good deletion.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java [hunk 4]
```diff
@@ -515,17 +510,15 @@ protected boolean processHandShake() {
         }
         if (numRefs != null) {
           if (acceptHandShake(endpointType, queueSize)) {
-            numRefs.incr();
+            ++numRefs;
```
Comment (bschuchardt) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java]: The old code updated a Counter held in the map.  This is just incrementing an Integer that you plucked out of the map.  It isn't going to update the value held in the map.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java [hunk 5]
```diff
@@ -539,14 +532,12 @@ protected boolean processHandShake() {
         return false;
       }
       synchronized (getCleanupProxyIdTable()) {
-        Counter numRefs = (Counter) getCleanupProxyIdTable().get(this.proxyId);
+        Integer numRefs = getCleanupProxyIdTable().get(this.proxyId);
         if (numRefs != null) {
-          numRefs.incr();
+          ++numRefs;
```
Comment (bschuchardt) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java]: same problem here - incrementing an integer isn't going to update the map

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerHandshakeProcessor.java [hunk 6]
```diff
@@ -71,7 +71,7 @@
 
   protected static final byte REPLY_INVALID = (byte) 61;
 
-  public static Version currentServerVersion = Acceptor.VERSION;
+  public static final Version currentServerVersion = Acceptor.VERSION;
```
Comment (bschuchardt) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerHandshakeProcessor.java]: I thought there were some tests that messed with this

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerHandshakeProcessor.java [hunk 7]
```diff
@@ -104,18 +94,4 @@ public static void refuse(OutputStream out, String message, byte exception) thro
     out.flush();
   }
 
-  // Keep the writeServerMember/readServerMember compatible with C++ native
-  // client
-  protected static void writeServerMember(DistributedMember member, DataOutputStream dos)
```
Comment (PivotalSarge) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerHandshakeProcessor.java]: Nice elision.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerHandshakeProcessor.java [hunk 8]
```diff
@@ -86,8 +74,10 @@ public static void refuse(OutputStream out, String message, byte exception) thro
 
     // Write the server's member
     DistributedMember member = InternalDistributedSystem.getAnyInstance().getDistributedMember();
-
-    writeServerMember(member, dos);
+    HeapDataOutputStream memberDos = new HeapDataOutputStream(Version.CURRENT);
+    DataSerializer.writeObject(member, memberDos);
```
Comment (bschuchardt) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerHandshakeProcessor.java]: Why did you remove the backward-compatibility serialization here?  If we make a change to the serialized form of this object won't this break backward-compatibility?

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImpl.java [hunk 9]
```diff
@@ -1493,6 +1500,43 @@ protected void handleNewClientConnection(final Socket socket,
...
+    dos.writeByte(exception);
+
+    // write dummy endpointType
+    dos.writeByte(0);
+    // write dummy queueSize
+    dos.writeInt(0);
+
+    // Write the server's member
+    DistributedMember member = InternalDistributedSystem.getAnyInstance().getDistributedMember();
+    HeapDataOutputStream memberDos = new HeapDataOutputStream(Version.CURRENT);
```
Comment (bschuchardt) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImpl.java]: Now this code has migrated to another class but I think it still has the same problem.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java (hunk 4 above)
Comment (WireBaron) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java]: Since this is using an Integer rather than an int, it's actually passing around a reference to the boxed value.  I was a little unsure about this myself, but verified in the debugger.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java (hunk 5 above)
Comment (WireBaron) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java]: I'll add some checks in the integration test to verify that we've properly incremented this state.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerHandshakeProcessor.java (hunk 6 above)
Comment (WireBaron) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerHandshakeProcessor.java]: Hmm, if so they must have been removed or modified.  This is building fine on both open and closed after this change.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerHandshakeProcessor.java (hunk 8 above)
Comment (WireBaron) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerHandshakeProcessor.java]: There shouldn't be any change in behavior from this refactor.  I just inlined the writeServerMember and cleaned up the code a bit as the output stream here can never be a VersionedDataStream (you can see where we construct the stream on line 66).  That being said, if there's a bug with the previous code not using the proper serialization here, let me know and we'll get that fixed.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImpl.java (hunk 9 above)
Comment (WireBaron) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImpl.java]: Let me know if there's a bug here.  The behavior shouldn't have changed, but I'm not entirely up to speed on the serialization code to see if there's an previous issue.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java (hunk 5 above)
Comment (bschuchardt) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java]: I wrote a simple test that showed the value in the map is unchanged.  With a Counter the value in the map would be incremented but it's not going to do that with a boxed integer.

map.put(key, 1);
Integer anInt = map.get(key);
anInt++;
Integer anotherInt = map.get(key);
System.out.println("anInt="+anInt+"; value in map="+anotherInt);

The result will be 2 and 1.

Comment (WireBaron) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java]: Thanks Bruce, I'm an idiot and managed to be testing the one case that wasn't actually incrementing the value (going from 0 to 1 and back).  Reverted this to use the counter again, which I'm not thrilled about, but seems less error prone than the other solutions I looked at.

Comment (PivotalSarge) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java]: Would it be clearer to rename it `MutableInteger`?
